### DISCUSSION TITLE
Return number of rows changed after flush

### DISCUSSION THREAD

**Discussion by mthaak** [sqlalchemy/sqlalchemy#6398] [Usage Questions]
**Is your feature request related to a problem? Please describe.**
I wanted to store a number of objects with `session.add_all(objects)` on a Postgres DB, where some objects could already exist in the database. Of course, this resulted in conflict errors. To get around this, I added this hook:
```python
@event.listens_for(engine, "before_cursor_execute", retval=True)
def _before_cursor_execute(conn, cursor, statement, parameters, context, executemany):
    if statement.startswith("INSERT "):
        statement = statement + " ON CONFLICT DO NOTHING"
    return statement, parameters
```
which skips over objects that already exist. 

Now I want to know how many rows have actually been updated in the database after the flush.

One way I found is doing:
```python
rowcount = 0
@event.listens_for(engine_survey, 'after_execute')
def receive_after_execute(conn, clauseelement, multiparams, params, execution_options, result):
    global rowcount
    rowcount += result.rowcount
```
which is of course very ugly.

So I would like an easy way to get the number of updated rows after a flush.

**Describe the solution you'd like**
I am not sure whether it makes sense, but what if `session.flush()` returns the number of rows that were changed during the flush.

**Additional context**
Using SQLAlchemy ORM

**Comment by zzzeek** [comment]
for any non-trivial application I would not recommend changing all INSERT statements across the board like that.

For PostgreSQL, we support upsert directly: https://docs.sqlalchemy.org/en/14/dialects/postgresql.html#insert-on-conflict-upsert  and you'd be best using this for your use case, although this does not integrate with the ORM right now.

as for session.flush(), this is not just a single UPDATE of a table, the flush could be doing any number of operations across any number of tables and there are also some cases where it emits UPDATE more than once on a row, so it's not like there's a simple formula for that.

There's things you can look at such as the length of the ``session.dirty`` list and such, however if you really want how many ORM updates it's doing I'd probably do the count you are doing, but use it inside a targeted hook like https://docs.sqlalchemy.org/en/14/orm/events.html?highlight=after_update#sqlalchemy.orm.MapperEvents.after_update .    You can store the count in session.info:
```python

import collections
from sqlalchemy import inspect

@event.listens_for(Session, 'before_flush')
def receive_before_flush(session, flush_context, instances):
    session.info['update_count'] = collections.defaultdict(int)

@event.listens_for(SomeClass, 'after_update')
def receive_after_update(mapper, connection, target):
    "listen for the 'after_update' event"
    inspect(target).session.info['update_count'][mapper.class_] += 1

```
then your counts are in session.info['update_count']

**Reply by mthaak** [reply]
Thanks @zzzeek for the suggestions.

I know about `Insert.on_conflict_do_nothing()` but since I was working with ORM, I wanted to stick with those. Converting the objects to plain python dicts wasn't clear to me (especially since the objects are nested with relationships).

Don't worry about the INSERTS being changed everywhere. My use case is a one-time-use script for migrating some data between two databases.

What I have now works fine 👌 . But I was just surprised that the way to do it requires all these hooks. And thought there must be an easier way.

**Reply by zzzeek** [reply]
it's not like it's a simple UPDATE of some rows, nobody has ever asked for this before and I can see how people would want it done in different ways, like do you want the sum of all UPDATES, or broken up by class, do you want to break it up by individual sub-tables for a class that's mapped to more than one table, not simple at all.   we give you the hooks to go in there instead.
